### ISSUE TITLE
Convert Plug proofs to 1st order rep

### ISSUE THREAD

**Issue by joaopizani** [joaopizani/piware-agda#27]
_From @joaopizani on March 28, 2015 12:32_

Until recently, Plug functions ("rewiring functions") were represented simply as `Fin o → Fin i`. However this representation made proving some of the necessary lemmas involving plugs complicated.

We suspect that a first-order representation as `Vec (Fin i) o` will be easier to manipulate and reason about. The substitution of one type itself for the other already happened, but now we need to **re-prove** the basic statements with plugs, given the new representation. This **re-proving** has 2 parts:
1. With _minimal effort_: use `tabulate` and so forth to prove the needed lemmas without changing much of the current proofs.
2. Try to actually **reorganize** the proofs for using facts about the different underlying type (vectors), and thus make the proofs simpler.

_Copied from original issue: joaopizani/piware#27_

**Comment by yoricksijsling** [comment]
I was also struggling with this for some proofs. One problem with using `Vec (Fin i) o` is that a definition of composition is kind of weird. It would involve some `tabulate`'s and `lookup`'s which becomes a bit messy.

For stretching, i used this plug:

```
plug-FM : ∀ {i o} → Morphism (vec-functor i) (vec-functor o) → 𝐂 i o
plug-FM M = Plug (op M (allFin _))
```

Where `Morphism (vec-functor i) (vec-functor o)` reduces to:

```
record Vec-Morphism (i : Ix) (o : Ix) : Set where
  field
    op : {∀ X} → Vec X i → Vec X o
    op-map-commute : {∀ X Y} (f : X Y) →
      (w : Vec X i) → (op ∘ mapᵥ f) w ≡ (mapᵥ f ∘ op) w
```

So in essence it's just a function on Vec's. The extra proof ensures that every element of the output vector must come from the input vector.

The reason i'm using this, is that `plug-FM` has the property that `⟦ plug-FM M ⟧ w ≡ op M w`. We don't have to mess about with `tabulate`'s and stuff. This is very useful for some proofs.

If we define some composition operator for morphisms `FM-∘`, we can easily derive that `plug-FM M₁ ⟫ plug-FM M₂ ≈⟦⟧ plug-FM (FM-∘ M₂ M₁)`.

Now i won't say that this definition is a good fit for all purposes. But it does make proofs with composed plugs a lot easier so i think it's useful to take a look at this.

**Comment by yoricksijsling** [comment]
So i was discussing this with Wouter yesterday.

The `map` and `op` are parametrically polymorphic. Simply by the types, we can assume that `op` and `map` _must_ commute. This is shown by Philip Wadler in Theorems for free.

The only downside is that the type theory of agda is not strong enough to figure this out, and we still have to prove it. But at least it's nice to know that the proof is always possible :).

Oh, also see [this answer on stack overflow](http://stackoverflow.com/questions/24718567/is-it-possible-to-get-hold-of-free-theorems-as-propositional-equalities).

**Comment by joaopizani** [comment]
Thx Yorick for the comments and replies. This issue was therefore closed in 58f41a23399c15aa16b89f0c1fa1638e53744ab5
